code/modules/logging/log_holder.dm

```diff
@@ -14,51 +11,143 @@ GLOBAL_PROTECT(logger)
+
+	/// list of Log args waiting for processing pending log initialization
+	var/list/waiting_log_calls
 
 	var/initialized = FALSE
 	var/shutdown = FALSE
 
+GENERAL_PROTECT_DATUM(/datum/log_holder)
+
 /// Assembles basic information for logging, creating the log category datums and checking for config flags as required
 /datum/log_holder/proc/init_logging()
 	if(initialized)
 		CRASH("Attempted to call init_logging twice!")
-	initialized = TRUE
 
 	round_id = GLOB.round_id
 	logging_start_timestamp = rustg_unix_timestamp()
 	log_categories = list()
 	disabled_categories = list()
-	for(var/datum/log_category/category_type as anything in subtypesof(/datum/log_category))
-		var/category = initial(category_type.category)
-		if(!category)
-			continue
 
-		if(category in log_categories)
-			stack_trace("Found two identical log category type definitions! [category_type]")
-			continue
+	category_group_tree = assemble_log_category_tree()
+	var/config_flag
+	for(var/datum/log_category/master_category as anything in category_group_tree)
+		var/list/sub_categories = category_group_tree[master_category]
+		for(var/datum/log_category/sub_category as anything in sub_categories)
+			config_flag = initial(sub_category.config_flag)
+			if(config_flag && !config.Get(config_flag))
+				disabled_categories[initial(sub_category.category)] = TRUE
+				sub_categories -= sub_category
+				continue
 
-		var/config_flag = initial(category_type.config_flag)
+		config_flag = initial(master_category.config_flag)
 		if(config_flag && !config.Get(config_flag))
-			disabled_categories[category] = TRUE
-			continue
-		category_type = log_categories[category] = new category_type
-		var/list/log_start_entry = list(
-			LOG_HEADER_CATEGORY = category,
-			LOG_HEADER_INIT_TIMESTAMP = big_number_to_text(logging_start_timestamp),
-			LOG_HEADER_ROUND_ID = big_number_to_text(GLOB.round_id),
-		)
-		rustg_file_write("[json_encode(log_start_entry)]\n", category_type.get_output_file(null))
+			disabled_categories[initial(master_category.category)] = TRUE
+			if(!length(sub_categories))
+				continue
+		// enabled, or any of the sub categories are enabled
+		init_log_category(master_category, sub_categories)
+
+	initialized = TRUE
+
+	// process any waiting log calls and then cut the list
+	for(var/list/arg_list as anything in waiting_log_calls)
+		Log(arglist(arg_list))
+	waiting_log_calls?.Cut()
 
 /// Tells the log_holder to not allow any more logging to be done, and dumps all categories to their json file
 /datum/log_holder/proc/shutdown_logging()
 	if(shutdown)
 		CRASH("Attempted to call shutdown_logging twice!")
 	shutdown = TRUE
 
-/// This is Log because log is a byond internal proc
+/// Iterates over all log category types to assemble them into a tree of main category -> (sub category)[] while also checking for loops and sanity errors
+/datum/log_holder/proc/assemble_log_category_tree()
+	var/static/list/category_tree
+	if(category_tree)
+		return category_tree.Copy()
```

Comment:
Why are we copying here? If it is needed, it should be done by consumers and not this API